python/ray/dag/compiled_dag_node.py

```diff
@@ -150,54 +138,58 @@ def _wrap_exception(exc):
+        NCCL write nodes. In this case, select the one that is the head of the
+        priority queue and its downstream nodes, regardless of whether the
+        downstream nodes are heads of their priority queues or not.
+
+    This function may return multiple nodes if they are NCCL nodes. In that case,
+    this function only removes the NCCL write node, which is also the head of a
+    priority queue. Other nodes will be removed in the following iterations.
+    Additionally, visited_nodes ensures that the same node will not be scheduled
+    more than once.
+
     Args:
-        task: The task to execute.
-        idx: The index of the task in the list of tasks of the actor.
-    Returns:
-        True if we are done executing all tasks of this actor, False otherwise.
+        actor_to_candidates: A dictionary mapping an actor handle to a list of
+            candidate nodes with zero in-degree.
+        graph: A dictionary mapping the index of a task to a dictionary of its
+            DAGOperationGraphNodes for different operations.
     """
-    # TODO: for cases where output is passed as input to a task on
-    # the same actor, introduce a "IntraProcessChannel" to avoid the overhead
-    # of serialization/deserialization and synchronization.
-    method = getattr(self, task.method_name)
-    input_reader = self._input_readers[idx]
-    output_writer = self._output_writers[idx]
-    res = None
-    try:
-        res = input_reader.read()
-    except RayChannelError:
-        # Channel closed. Exit the loop.
-        return True
-
-    try:
-        _process_return_vals(res, return_single_output=False)
-    except Exception as exc:
-        # Previous task raised an application-level exception.
-        # Propagate it and skip the actual task. We don't need to wrap the
-        # exception in a RayTaskError here because it has already been wrapped
-        # by the previous task.
-        output_writer.write(exc)
-        return False
-
-    resolved_inputs = []
-    for task_input in task.task_inputs:
-        resolved_inputs.append(task_input.resolve(res))
-
-    try:
-        output_val = method(*resolved_inputs, **task.resolved_kwargs)
-    except Exception as exc:
-        output_val = _wrap_exception(exc)
-
-    try:
-        output_writer.write(output_val)
-    except RayChannelError:
-        # Channel closed. Exit the loop.
-        return True
-
-    return False
+    next_nodes = []
+    first_nccl_node = None
+    for _, candidates in actor_to_candidates.items():
+        if (
+            not candidates[0].requires_nccl
+            or candidates[0].operation.type != DAGNodeOperationType.WRITE
+        ):
+            next_nodes.append(heapq.heappop(candidates))
+            return next_nodes
+        if first_nccl_node is None:
+            first_nccl_node = candidates[0]
+
+    assert first_nccl_node is not None
+    next_nodes.append(heapq.heappop(actor_to_candidates[first_nccl_node.actor_handle]))
```

Comment:
this is supposed to be READ? can you add assert?

Reply:
It should be `WRITE`. Updated https://github.com/ray-project/ray/pull/46911/commits/c1becc9af8ca1cc3ae5bf8cd21029fb7cd64e16b